src/source/geojson_worker_source.js

```diff
@@ -293,4 +296,70 @@ class GeoJSONWorkerSource extends VectorTileWorkerSource {
     }
 }
 
+function getSuperclusterOptions({superclusterOptions, clusterProperties}) {
+    if (!clusterProperties || !superclusterOptions) return superclusterOptions;
+
+    const initialValues = {};
+    const mapExpressions = {};
+    const reduceExpressions = {};
+    const globals = {accumulated: null, zoom: 0};
+    const feature = {properties: null};
+    const propertyNames = Object.keys(clusterProperties);
+
+    for (const key of propertyNames) {
+        const [operator, initialExpression, mapExpression] = clusterProperties[key];
+
+        const parsed = createExpression(clusterProperties[key]);
+        if (parsed.result === 'error') {
+            const message = parsed.value.map(e => e.message).join('; ');
+            throw new Error(`Error parsing expression for cluster property "${key}": ${message}`);
+
+        } else if (!isGlobalPropertyConstant(parsed.value.expression, ['zoom'])) {
+            throw new Error(`Error parsing expression for cluster property "${key}": zoom expressions not supported.`);
+        }
+
+        const initialExpressionParsed = createExpression(initialExpression);
+        const mapExpressionParsed = createExpression(mapExpression);
+        const reduceExpressionParsed = createExpression([operator, ['accumulated'], ['get', key]]);
+
+        if (initialExpressionParsed.result === 'success') {
+            if (!isFeatureConstant(initialExpressionParsed.value.expression))
+                throw new Error(`Error parsing expression for cluster property "${key}": can't use feature properties in initial expression.`);
+
+            initialValues[key] = initialExpressionParsed.value.evaluate(globals);
+        }
+
+        if (mapExpressionParsed.result === 'success')
+            mapExpressions[key] = mapExpressionParsed.value;
+
+        if (reduceExpressionParsed.result === 'success')
+            reduceExpressions[key] = reduceExpressionParsed.value;
+    }
+
+    superclusterOptions.initial = () => {
+        const properties = {};
+        for (const key of propertyNames) {
+            properties[key] = initialValues[key];
+        }
+        return properties;
+    };
+    superclusterOptions.map = (pointProperties) => {
+        feature.properties = pointProperties;
```

Comment:
Why is this (and in `reduce below`) using a shared `feature` and overriding the `properties` of it instead of just defining a new `const feature = { properties: pointProperties }` here?

Reply:
The map function and expression evaluation is synchronous, meaning there are no side effects from using mutable state here (which is reset at the beginning). At the same time, we avoid constantly creating new feature objects, improving performance and lots of garbage collection passes.

Reply:
Ah, OK that makes sense. I thought it might have something to do with performance optimization.